pkg/client/william.go

```diff
@@ -0,0 +1,233 @@
+package client
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"net/url"
+	"strings"
+	"sync"
+)
+
+// William Interface
+type William interface {
+	ByPass()
+	SetClient(client HttpClient)
+	SetKeyPair(id, secret string)
+	GetServiceName() string
+	ListPermission(ctx context.Context, ownershipLevel, action string, resourceHierarchy ...string) ([]byte, error)
+	HandlerFunc(permission func(r *http.Request) string, next http.HandlerFunc) http.HandlerFunc
+	AmHandler(w http.ResponseWriter, r *http.Request)
+	AddAction(action string)
+	AddActionFunc(action string, f ActionResourceFunc)
+	Generate(ownershipLevel, action string, resourceHierarchy ...string) func(r *http.Request) string
+}
+
+type AmPermission struct {
+	Alias    string `json:"alias,omitempty"`
+	Prefix   string `json:"prefix"`
+	Complete bool   `json:"complete"`
+}
+
+type ActionResourceFunc func(context.Context, string) []AmPermission
+
+type HttpClient interface {
+	Do(req *http.Request) (*http.Response, error)
+}
+
+type william struct {
+	sync.RWMutex
+	actions map[string]ActionResourceFunc
+
+	keyID     string
+	keySecret string
+
+	serviceName string
+	baseURL     string
+	bypass      bool
+
+	client HttpClient
+}
+
+// New create a new william client
+func New(baseURL, serviceName string) William {
```

Comment:
Since we're in the package `client`, and this is its `New` function, I'd expect its return to be a client, not `William`. It's usual to shorten the name of the "main" type of the package, so it doesn't stutter. Eg: `New() *C`, so we get a `*client.C` instead of a `*client.Client`.